[Backtracing][Linux] Add Linux support to swift-backtrace.

al45tair opened the pull request:
We need a Linux specific `Target` implementation, and a couple of minor tweaks to make things build everywhere.

rdar://110262673

al45tair committed:
[Backtracing][IRGen] Add a semantic attribute to force frame pointer.

The Swift backtracer's frame pointer unwinder cannot work on Linux
without this change, because the compiler omits the frame pointer from
the function in libSwift_Backtracing that actually captures the stack.

rdar://110260855

al45tair committed:
[Backtracing][Linux] Reorganise modules for _Backtracing.

Using `SwiftShims` is undesirable - it creates all kinds of build issues,
and means shipping the `_SwiftBacktracing.h` header in the SDK, which is
not necessary.

While we're doing this, add the necessary definitions for reading ELF
and DWARF information.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Add Linux support to the _Backtracing module.

Use the new module structure rather the old SwiftShims header.  This
is much cleaner and lets us include operating system headers to get
the relevant definitions where possible.

Add code to support ELF and DWARF, including decompression using
zlib, zstd and liblzma if those turn out to be required and available.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Add Linux support to swift-backtrace.

We need a Linux specific `Target` implementation, and a couple of minor
tweaks to make things build everywhere.

rdar://110262673

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing][Linux] Temporarily add a default addressWidth.

This is for compatibility, so that I can split up the PRs.
We'll remove it in the next PR.

rdar://110261712

al45tair committed:
Revert "[Backtracing][Linux] Temporarily add a default addressWidth."

This reverts commit 0d7c45d9dac68ca820baec816b8857070168136a.

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing][Linux][Tests] Fix verify_all_overlays test.

The `_Backtracing` module has a number of private implementation only
imports that aren't used outside of the module and that don't require
any additional libraries (hence they aren't relevant to the outside
world).  `verify_all_overlays.py` needs to know about these when it
does its test, because it loadas the module as the main module, which
results in implementation only imports being required instead of
ignored.

rdar://110261712

al45tair commented:
When reviewing, you can start at 450011e. Previous changesets are being reviewed in #66333, #66334 and #66335.

al45tair commented:
If you're wondering where the tests are, I left them until the last PR (#66338) so everything was enabled and they'd actually work.

MaxDesiatov commented on stdlib/public/libexec/swift-backtrace/TargetLinux.swift:
```diff
@@ -0,0 +1,232 @@
+    }
+  }
+
+  public func redoBacktraces(limit: Int?, top: Int, cache: Bool) {
```
IIUC, the type it's declared on is `internal`, do these `public` access modifiers have any effect here?

al45tair replied:
No. Removed.

MaxDesiatov commented on stdlib/public/libexec/swift-backtrace/TargetLinux.swift:
```diff
@@ -0,0 +1,232 @@
+  /// Fetch information about all of the process's threads; the crash_info
+  /// structure contains a linked list of thread ucontexts, which may not
+  /// include every thread.  In particular, if a thread was stuck in an
+  /// uninterruptible wait, we won't have a ucontext for it.
```
Could more context be added for each parameter in this doc comment?

al45tair replied:
`limit` and `top` are the same as elsewhere, and I think the list head argument is obvious. I'm not inclined to update the comment here.

al45tair committed:
[Backtracing][Linux] Work correctly in the presence of partial reads.

There's a chance that pipes might perform a partial read; we should
handle that case.

rdar://110261712

al45tair commented on stdlib/public/libexec/swift-backtrace/TargetLinux.swift:
```diff
@@ -0,0 +1,232 @@
+    return trimmed
+  }
+
+  init(crashInfoAddr: UInt64, limit: Int?, top: Int, cache: Bool) {
```
`limit` and `top` are the same as the parameters documented in the `_Backtracing` module's interface. I don't think documenting them twice is a good idea, since the documentation could then diverge. See the doc comment for `Backtrace.capture()`, as well as the comment for the `Frames.omittedFrames(count)` case

MaxDesiatov replied:
Right, then a doc comment for this initializer could reference the source of truth. Otherwise there's nothing here and no mention of the place where this information could be looked up. Some redirection would be helpful to provide the context.

al45tair committed:
[Backtracing][Linux] Remove some stray "public" annotations.

We don't need these here.  They were a leftover from previous code.

rdar://110262673

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing][Linux] Remove status argument.

The `status` argument to the `_swift_backtrace_demangle()` function
isn't especially useful, won't match behaviour on Windows, and we
actually don't use it in the Swift code that calls this SPI.

Remove it.

rdar://110261712

al45tair committed:
[Backtracing][Linux] Add a comment about __cxa_demangle.

`__cxa_demangle()` is a rather unusual API; one of its "features" is that
the pointer you pass in must either be `nullptr`, in which case it
will call `malloc()` itself, _or_ it has to be a pointer to a block
of memory allocated with `malloc()`, because `__cxa_demangle()` may
`realloc()` it for you.

This seems to me to be something of a non-obvious footgun, so we never
pass the caller's pointer through to `__cxa_demangle()`, which lets them
decide how they want to allocate space.

rdar://110261712

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Backtracing][Linux] Also remove the status argument from Runtime module.

There's a separate declaration here because we can't include the `Backtrace.h`
header from inside `modules/Runtime/Runtime.h`.

rdar://110261712

al45tair commented:
@swift-ci Please smoke test
